lib/kube/proxy/transport_retryable.go

```diff
@@ -0,0 +1,191 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"log/slog"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/semaphore"
+)
+
+// retryableTransport wraps an http.RoundTripper to enable automatic retry
+// on HTTP/2 GOAWAY frames by buffering request bodies in memory.
+//
+// Kubernetes API servers send GOAWAY on HTTP/2 connections to redistribute
+// load across replicas. Go's http2.Transport automatically retries if
+// req.GetBody is set.
+//
+// A global semaphore limits total buffered memory to 500 MiB across all
+// concurrent requests, preventing OOM. Individual requests have a maximum
+// body size of 50 MiB. Requests are blocked if required memory cannot be
+// acquired from the semaphore.
+//
+// Requests that are not made retryable are:
+// 1. HTTP/1.1 protocol upgrades (exec/attach/portforward), which won't
+// receive an HTTP/2.0 GOAWAY.
+// 2. Requests with body sizes greater than 50 MiB.
+// 3. Requests with unknown body sizes used with chunked encoding
+// cannot safely acquire the semaphore.
+type retryableTransport struct {
+	inner     http.RoundTripper
+	log       *slog.Logger
+	semaphore *semaphore.Weighted
+}
+
+const (
+	// maxTotalBufferSize limits total memory used for buffering request bodies
+	// across all concurrent requests to prevent OOM at scale.
+	maxTotalBufferSize = 500 * 1024 * 1024 // 500 MiB
+
+	// maxBufferPerRequest limits individual request body size to allow
+	// balanced memory usage across requests. Kubernetes ConfigMaps and Secrets
+	// have 1 MiB etcd limits, so 50 MiB is generous.
+	maxBufferPerRequest = 50 * 1024 * 1024 // 50 MiB
+)
+
+var retryBufferSemaphore = semaphore.NewWeighted(maxTotalBufferSize)
```

Comment:
Having this in a global means that testing is awkward (since different instances of TeleportProcess in tests will share the same semaphore) and exposing this as a tunable (which we should definitely do, even if with a `TELEPORT_UNSTABLE_` envvar) is basically impossible to do sanely.

The buffer semaphore should be owned by `Forwarder` and passed to `newRetryableTransport`, I think.